Dotnetup: Shell profile modification for PATH and DOTNET_ROOT (#53447)
dotnet/sdk: documentation/general/dotnetup/unix-environment-setup.md

## Overview

> dotnetup automatically configures the Unix shell environment so that .NET is available in every new terminal session. This involves modifying shell profile files to set the `PATH` and `DOTNET_ROOT` environment variables. The same mechanism also supports PowerShell on any platform.

Reviewer: ```suggestion
dotnetup (upon user consent) configures the Unix shell environment so that .NET is available in every new terminal session. This involves modifying shell profile files to set the `PATH` and `DOTNET_ROOT` environment variables. The same mechanism also supports PowerShell on any platform.
```

Reviewer: nit: we aren't supposed to do this without consent, but it is the default

Author: Let's go with "dotnetup can configure".

> On Windows the primary method is registry-based environment variables, which is handled separately. This document focuses on the Unix (and PowerShell) profile-based approach.

Reviewer: ```suggestion
This document focuses on the Unix (and PowerShell) profile-based approach. On Windows, registry-based environment variables also impact the environment in addition to profile files.
```

  ```
  To start using .NET in this terminal, run:
    eval "$('/home/user/.local/share/dotnetup/dotnetup' print-env-script --shell bash)"
  ```

If the user already has a saved path preference, or if the command is non-interactive / uses an explicit `--install-path`, the walkthrough prompt is skipped and dotnetup uses the existing configuration or the explicit path directly.

### 2. `dotnetup defaultinstall`

A standalone command that explicitly configures (or reconfigures) the default .NET install:

```bash
# Set up user-level default install (modifies shell profiles)
dotnetup defaultinstall user

# Switch to admin/system-managed .NET (removes DOTNET_ROOT from profiles, keeps dotnetup on PATH)
dotnetup defaultinstall admin
```

**`defaultinstall user`** on Unix:
1. Detects the current shell
2. Modifies the appropriate shell profile files
3. Prints an activation command for the current terminal

**`defaultinstall admin`** on Unix:
- Replaces existing profile entries with dotnetup-only entries (keeps dotnetup on PATH but removes `DOTNET_ROOT` and dotnet from `PATH`), since the system package manager owns the .NET installation.

## Shell Profile Modification

> ### Which Profile Files Are Modified

Reviewer: I think this section will likely need to be updated

Author: How so?

Reviewer: To include the PROFILE and ZSH env variable  or other lookup approaches!

| Shell | Files modified | Rationale |
|-------|---------------|-----------|
| **bash** | `~/.bashrc` (always) + the first existing of `~/.bash_profile` / `~/.profile` (creates `~/.profile` if neither exists) | `.bashrc` covers Linux terminals (non-login shells). The login profile covers macOS Terminal and SSH sessions. We never create `~/.bash_profile` to avoid shadowing an existing `~/.profile`. |
| **zsh** | `~/.zshrc` (created if needed) | Covers all interactive zsh sessions. `~/.zshenv` is avoided because on macOS, `/etc/zprofile` runs `path_helper` which resets PATH after `.zshenv` loads. |
| **pwsh** | `~/.config/powershell/Microsoft.PowerShell_profile.ps1` (creates directory and file if needed) | Standard `$PROFILE` path on Unix. |

### Profile Entry Format

Each profile file gets a marker comment and an eval line:

**Bash / Zsh:**
```bash
# dotnetup

> eval "$('/path/to/dotnetup' print-env-script --shell bash)"

Reviewer: I don't think `eval` is the best approach here, though I have limited knowledge. 

`eval` will fail if dotnetup get uninstalled, and I don't know which shells can recover from a failed eval line. 

I think we should be sourcing a file that we write instead, and we should make sure if the file no longer exists we print a message but ignore and don't fail the rest of the profile from loading. `eval` invokes a process and running an executable will add overhead to every shell launch. I believe sourcing files is what other tools like rustup do as well. 

I see the benefit of this in theory if the dotnet directory changes, then we would always regen the script with  the correct, new directory, but also if dotnetup's directory changes this would not work. (of course a sourced .dotnetup  file would also not be found if dotnetup is moved but that's ok)

Author: @baronfel What do you think?  I think we landed on eval because it's a lot simpler in some ways.  But do we need to worry about what happens if the dotnetup it points to no longer exists?  How should we handle that?

Reviewer 2: I'm not too worried about this failure mode - but we _can_ be defensive here. We can test if the binary exists very easily and only eval from it if it does.

```bash
if command -v dotnetup &>/dev/null
then
  # eval
else
  # swallow or print to stderr that dotnetup is missing
fi
```

Reviewer: Do you lean towards eval rather than sourcing the files? I think this is different from what the other tools do, but admittedly we have more challenges with VS + the muxer.

I didn't do the performance metrics on this, but from my experience trying to run a process in a terminal, I anecdotally expect this to add anywhere between a 50-400ms delay to the profile launch. Users aren't going to have the context/nuance of the eval vs .source approach, I certainly wouldn't be happy if installing rustup made my terminal sessions startup a lot slower because it was running rustup every time. For users who don't do much dotnet dev this might be a pain point for them... but I'm happy if I'm wrong

Reviewer 2: >50-400ms

Why is this the case? a NAOT app running static only code should be in the <10ms space.

Reviewer: This is what I observed. I think there's overhead with any potential telemetry,  sign check, (windows defender?), overhead launching an executable, etc.
```
$sw = [System.Diagnostics.Stopwatch]::StartNew(); & 'C:\Users\username\dotnetup.exe' print-env-script --shell pwsh | Out-Null; $sw.Stop(); Write-Host "$($sw.ElapsedMilliseconds) ms"
```

```
Iteration | Time
1 | 218 ms
2 | 97 ms
3 | 114 ms

Average | 143 ms
```

To make sure it's not the powershell calls, I also tried this:
```
# Measure raw process execution time without PowerShell pipeline overhead
$psi = [System.Diagnostics.ProcessStartInfo]::new('C:\Users\username\dotnetup.exe', 'print-env-script --shell pwsh')
$psi.UseShellExecute = $false
$psi.RedirectStandardOutput = $true
$psi.RedirectStandardError = $true
$sw = [System.Diagnostics.Stopwatch]::StartNew()
$proc = [System.Diagnostics.Process]::Start($psi)
$stdout = $proc.StandardOutput.ReadToEnd()
$proc.WaitForExit()
$sw.Stop()
Write-Host "Raw process time: $($sw.ElapsedMilliseconds) ms"
Write-Host "Exit code: $($proc.ExitCode)"
Write-Host "Output lines: $(($stdout -split "`n").Count)"
Write-Host "---Output---"
$stdout
```

and it averaged 196 ms for me

Reviewer: I wonder if we should try to disable telemetry in the dotnetup invocation and if that'd help - obviously we wouldn't get the info on eval though. We could do a follow up issue to investigate this, don't need to block the PR on this.

Author: I filed #54029 to follow up on the run speed in general.

Reviewer: Thanks - resolving!

```

**PowerShell:**
```powershell
# dotnetup

> & '/path/to/dotnetup' print-env-script --shell pwsh | Invoke-Expression

Reviewer: `| Invoke-Expression` - this fails for me when I tried to run it : `Invoke-Expression: Cannot bind argument to parameter 'Command' because it is an empty string.` - I don't think we need invoke-expression - we may need to update the code itself if it does this.

Author: This should be fixed.  You do need `Invoke-Expression`, but you need `Out-String` to avoid it passing each line separately.  An empty line was causing the error.

> 

Reviewer: ```suggestion

Supported shell values are based on what `nvm` and `rustup` support today.
```

We could document our decision making here.
